chain/network/src/peer_manager.rs

```diff
@@ -371,8 +380,25 @@ impl PeerManagerActor {
         });
     }
 
+    fn num_active_outgoing_peers(&self) -> usize {
+        self.active_peers
+            .values()
+            .filter(|active_peer| active_peer.peer_type == PeerType::Outbound)
+            .count()
+    }
+
+    /// Check if it is needed to create a new outbound connection.
+    /// If the number of active connections is less than `ideal_connections_lo` or
+    /// (the number of outgoing connections is less than `minimum_outbound_peers`
+    ///     and the total connections is less than `max_peers`)
     fn is_outbound_bootstrap_needed(&self) -> bool {
-        self.active_peers.len() + self.outgoing_peers.len() < self.config.max_peer as usize
+        let total_connections = self.active_peers.len() + self.outgoing_peers.len();
+        let potential_outgoing_connections =
+            self.num_active_outgoing_peers() + self.outgoing_peers.len();
+
+        (total_connections < self.config.ideal_connections_lo as usize
+            || total_connections < self.config.max_peer as usize
+                && potential_outgoing_connections < self.config.minimum_outbound_peers as usize)
```

Comment:
This means that as long as the number of outgoing connections is smaller than `minimum_outbound_peers`, we will initiate new outbound connections regardless of the total number of connections. Is it intended?

Reply:
😅 I missed some parenthesis here, nice catch!

Reply:
@bowenwang1996  Actually after checking [operation precedence](https://doc.rust-lang.org/reference/expressions.html) `&&` is evaluated before `||` so the statement is correct, however the missing parenthesis was not intentional. Check now, and let me know if it makes sense.